packages/hardhat/contracts/Pools.sol

```diff
@@ -0,0 +1,161 @@
+    }
+
+    function balanceOf(address token, address owner) external view returns (uint) {
+        Pool storage pool = pools[token];
+        return pool.totalShare > 0
+        ? PRBMath.mulDiv(pool.balance, pool.shares[owner], pool.totalShare)
+        : 0;
+    }
+
+    function shareOf(address token, address owner) external view returns (uint) {
+        Pool storage pool = pools[token];
+        return pool.totalShare > 0
+        ? PRBMath.mulDiv(100, pool.shares[owner], pool.totalShare)
+        : 0;
+    }
+
+    function deposit(address token, uint256 amount) external validUAmount(amount) returns (uint256 share) {
+
+        Pool storage pool = pools[token];
+
+        share = pool.totalShare > 0
+        ? PRBMath.mulDiv(amount, pool.totalShare, pool.balance)
+        : amount * (10 ** ERC20(token).decimals());
+
+        pool.balance = pool.balance + amount;
+        pool.totalShare = pool.totalShare + share;
+        pool.shares[msg.sender] = pool.shares[msg.sender] + share;
+
+        ERC20(token).safeTransferFrom(msg.sender, address(this), amount);
+
+    }
+
+    function withdraw(address token, uint amount) external validUAmount(amount) {
+        Pool storage pool = pools[token];
+
+        require(pool.balance >= amount, "Amount too big");
+
+        uint256 share = PRBMath.mulDiv(pool.totalShare, amount, pool.balance);
+
+        require(pool.shares[msg.sender] >= share, "Amount gt than balance");
+
+        pool.balance = pool.balance - amount;
+        pool.totalShare = pool.totalShare - share;
+        pool.shares[msg.sender] = pool.shares[msg.sender] - share;
+
+        ERC20(token).safeTransfer(msg.sender, amount);
+
+    }
+
+    function borrow(address token, uint amount, address recipient) external override validUAmount(amount) validAddress(recipient) {
+        Pool storage pool = pools[token];
+        pool.borrowed = pool.borrowed + amount;
+        ERC20(token).safeTransfer(recipient, amount);
+    }
+
+    function repay(address token, uint amount, uint interest) external override validUAmount(amount) validUAmount(interest) {
+        Pool storage pool = pools[token];
+        require(pool.borrowed >= amount, "Amount too big");
+        pool.balance = pool.balance + interest;
+        pool.borrowed = pool.borrowed - amount;
+    }
+
+    function borrowingRate(address token) external view override returns (uint rate) {
+        rate = _borrowingRateAfterLoan(token, 0);
+    }
+
+    function borrowingRateAfterLoan(address token, uint amount) validUAmount(amount) external view override returns (uint rate) {
+        rate = _borrowingRateAfterLoan(token, amount);
+    }
+
+    function utilisationRate(address token) external view override returns (uint rate) {
+        rate = _utilisationRateAfterLoan(token, 0);
+    }
+
+    function _borrowingRateAfterLoan(address token, uint amount) internal view returns (uint rate) {
+        Pool storage pool = pools[token];
```

Comment:
I think you want to make it `memory` here, as you're not writing, something tells me it's cheaper, but we should confirm